File: dynamic-programming/beecrowd-1288/solution.py
from sys import stdin


def knapsack(power: list[int], weight: list[int], capacity: int) -> int:
    n = len(power)
    dp = [[0] * (capacity + 1) for _ in range(n + 1)]

    for i, wt in enumerate(weight, 1):
        for j in range(1, capacity + 1):
            if wt <= j:
                dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - wt] + power[i - 1])
            else:
                dp[i][j] = dp[i - 1][j]

    return dp[n][capacity]


def main():
    T = int(stdin.readline())

    for _ in range(T):
        n = int(stdin.readline())
        power, weight = [0] * n, [0] * n

        for i in range(n):
            power[i], weight[i] = map(int, stdin.readline().split())

        k, r = int(stdin.readline()), int(stdin.readline())

        print(
            "Falha na missao"
            if knapsack(power, weight, k) < r
            else "Missao completada com sucesso"
        )


if __name__ == "__main__":
    main()

File: dynamic-programming/beecrowd-1288/solution.cpp
#include <cstdio>
#define max(x, y) ((x) > (y) ? (x) : (y))
#define size 101
using namespace std;

int knapsack(int W, int wt[], int val[], int n) {
  int M[size][size];
  for (int i = 0; i <= n; i++) {
    for (int w = 0; w <= W; w++) {
      if (i == 0 || w == 0)
        M[i][w] = 0;
      else if (wt[i - 1] <= w)
        M[i][w] = max(val[i - 1] + M[i - 1][w - wt[i - 1]], M[i - 1][w]);
      else
        M[i][w] = M[i - 1][w];
    }
  }
  return M[n][W];
}

int main() {
  int T, n, val[100], wt[100], k, r;
  scanf("%d", &T);
  while (T--) {
    scanf("%d", &n);
    for (int i = 0; i < n; i++)
      scanf("%d %d", &val[i], &wt[i]);
    scanf("%d", &k);
    scanf("%d", &r);
    printf("%s\n",
           (knapsack(k, wt, val, n) < r ? "Falha na missao"
                                        : "Missao completada com sucesso"));
  }

  return 0;
}
